### PR TITLE
exclude judges from Other attorney list in Cases to Assign page

### LINKED ISSUE #12982: Fix RecordInvalid: Validation failed: Assigned to has to be an attorney
## Description

When judge assigns case to user, validation error `Assigned to has to be an attorney` occurs. Judge should not be given an option to choose an invalid user (e.g., a non-attorney).

When `app/controllers/judge_assign_tasks_controller.rb in block (2 levels) in tasks at line 17` is called, `app/workflows/attorney_task_creator.rb in tasks at line 20` raises validation error: `ActiveRecord::RecordInvalid: Validation failed: Assigned to has to be an attorney`

## Background/context/resources
During Bat Teaming, got msg: https://dsva.slack.com/archives/C4JECDLSE/p1576502206014100

Sentry alerts: 
* https://sentry.ds.va.gov/department-of-veterans-affairs/caseflow/issues/7329/
* https://sentry.ds.va.gov/department-of-veterans-affairs/caseflow/issues/7547/

## AC
- [ ] Only attys are listed on the judge assign page
- [ ] Jim notified for release notes

### PR DESCRIPTION
**Pull request by yoomlam**
Resolves #12982

### Description
On the Assign Cases page (e.g., http://localhost:3000/queue/3/assign), when a judge chooses to Assign case(s) to "Other", a secondary dropdown selection field offers other attorneys for the judge to choose. This PR ensures no judges are included in the selection options because choosing a judge causes an error (in the UI and as a Sentry alert).

To reassign a task to a judge, the judge should go to the case details screen. This is the desired behavior based on this test: https://github.com/department-of-veterans-affairs/caseflow/blob/master/spec/feature/queue/judge_assignment_spec.rb#L278

### Acceptance Criteria
- [ ] No judges are listed in the secondary dropdown; acting judges should be on the list
- [ ] Case assignment to attorneys still work
- [ ] Re-assign to a judge on the Case Details page still works

### Testing Plan
1. Log in as a judge that has cases to assign (e.g., BVAAABSHIRE)
2. Go to the "Cases to Assign" page (e.g., http://localhost:3000/queue/3/assign)
3. Click on "Other" in the "Assign to:" dropdown
4. In the secondary dropdown, choose any user
5. Select cases for assignment
6. Click "Assign N cases" button
7. Re-assign a case to a judge on the Case Details page -- follow steps in the comment below with screenshots 

# User Facing Changes
 - [ ] Screenshots of UI changes added to PR & Original Issue

### Judges are listed in secondary dropdown (in this case the judge is named "Judge has case to assign"): 
image

### Error message after clicking "Assign 1 case" to that judge:
image

### Judges are NOT listed in secondary dropdown:
image

### Acting Judges are included in the dropdown:
image

### Assignment to Acting judge is successful
image

### Assignment to attorney on any team is successful
image

### CODE REVIEW

**Comment by yoomlam** [comment]
Actual counts from prod:
```
attys_and_judges = VACOLS::Staff.where(sactive: "A").where.not(sattyid: nil).where.not(sdomainid: nil)
true_judges = VACOLS::Staff.where(svlj: "J", sactive: "A").where.not(sdomainid: nil)
attys_and_judges.size
=> 999
true_judges.size
=> 127
attys_and_acting_judges = attys_and_judges - true_judges
attys_and_acting_judges.size
=> 875
999 - 127
=> 872
```

Why the difference of 3?
```
huh = true_judges - attys_and_judges
huh.size
=> 3
huh.pluck(:sdomainid)
=> ["VACOBRAYW", "VACOHIGHTJ", "BVAFTR1"]
huh.pluck(:snamel)
=> ["Bray", "Hightower", "Recording"]
huh.pluck(:sattyid)
=> [nil, nil, nil]
```

File: app/repositories/attorney_repository.rb [hunk 1]
```diff
@@ -1,10 +1,21 @@
...
       .pluck("UPPER(sdomainid)")
 
     User.batch_find_by_css_id_or_create_with_default_station_id(css_ids)
   end
+
+  def self.find_all_attorneys_without_judges
+    # Optimize these into a single query once we agree this is what we want
+    attys_and_judges = VACOLS::Staff.where(sactive: "A").where.not(sattyid: nil).where.not(sdomainid: nil).pluck("UPPER(sdomainid)")
+    true_judges = VACOLS::Staff.where(svlj: "J", sactive: "A").where.not(sdomainid: nil).pluck("UPPER(sdomainid)")
+    attys_and_acting_judges = attys_and_judges - true_judges
```
Comment (hschallhorn) [inline app/repositories/attorney_repository.rb]: Don't think we need two calls to vacols here. `.where(sactive: "A").where.not(sattyid: nil)` ensures we get everyone that is an active attorney, judge, or acting judge. From there, we can just exclude "true judges" with `.where.not(svlj: "J")`.

I'd be interested to know if working this logic into `find_all_attorneys` has any unwanted side effects. I know we only pull attorneys  in the two loading screens (queue and case details) but if that's only done to populate this dropdown, we can just have `find_all_attorneys` return only attorneys.

Comment (yoomlam) [reply app/repositories/attorney_repository.rb]: Besides those 2 screens you mentioned, `find_all_attorneys` is used be the `UsersController`. Is there an easy way to check which client code makes request to the `/users` endpoint? My first approach would be grep.

Comment (hschallhorn) [reply app/repositories/attorney_repository.rb]: None great that I really know of. grepping `role=attorney` should yield some results

Comment (yoomlam) [reply app/repositories/attorney_repository.rb]: If I'm interpreting your suggestion for combining the 2 queries correctly, that will produce a *set intersection*.  My implementation is a *set difference*.

**Comment by yoomlam** [comment]
Here's some info to help clarify discussions.

All counts below are for active users with a CSS_ID (`VACOLS::Staff.where(sactive: "A").where.not(sdomainid: nil)`). Bolded numbers are sums for the row or column. 
Based on our current code, I'm interpreting a user with a non-nil `attyid` to be an attorney.

  |   | An attorney (has attyid) | Not attorney (attyid=nil) |  
-- | -- | -- | -- | --
Acting judge | svlj="A" | 187 | 71 | **258**
Judge | svlj="J" | 124 | 3 | **127**
  | svlj=nil | 688 | 346 | **1034**
  | svlj="I" | 0 | 0 | **0**
  |   | **999** | **420** |  

* [`AttorneyRepository.find_all_attorneys.size`](https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/repositories/attorney_repository.rb#L5) returns 999. The query result includes judges with non-nil `attyid`s.
* [`JudgeRepository.find_all_judges.size`](https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/repositories/judge_repository.rb#L24) returns 385 (or **258** + **127** from the table).
* For this ticket, I believe we want to show attorneys and acting judges for the user to choose:
999 attorneys - 124 judges with attyid = 875. Is this correct?

Surprises:
* There are 71 acting judges who are not attorneys (or at least don't have an `attid`).  
  An acting judge is defined as an attorney who has been given a judge role (temporarily).

**Comment by yoomlam** [comment]
> Release notes to remind users they can only assign to attorneys from the assign page and to reassign to another judge from the case details screen (include screenshots).

### Case details page with the option to "Re-assign to a judge":
image

### Pop-up window to select a judge:
image

### Selection and notes before submitting:
image

### After submitting:
image

File: app/repositories/attorney_repository.rb [hunk 2]
```diff
@@ -1,10 +1,19 @@
 # frozen_string_literal: true
 
 class AttorneyRepository
-  def self.find_all_attorneys
+  # this includes judges
+  def self.find_all_having_attorney_ids
```
Comment (hschallhorn) [inline app/repositories/attorney_repository.rb]: Is this used anywhere?

File: app/repositories/judge_repository.rb [hunk 3]
```diff
@@ -20,8 +20,19 @@ def self.find_all_judges_with_name_and_id
...
+
+  # :nocov:
+  # excludes acting judges
+  def self.judges
+    VACOLS::Staff.where(svlj: "J", sactive: "A")
+  end
+
+  def self.acting_judges
+    VACOLS::Staff.where(svlj: "A", sactive: "A")
+  end
```
Comment (hschallhorn) [inline app/repositories/judge_repository.rb]: Similarly, I'm not seeing these being used anywhere

File: app/models/attorney.rb [hunk 4]
```diff
@@ -2,9 +2,9 @@
 
 class Attorney
   class << self
-    def list_all
-      Rails.cache.fetch("#{Rails.env}_list_of_attorneys_from_vacols") do
-        AttorneyRepository.find_all_attorneys
+    def list_all_excluding_judges
```
Comment (hschallhorn) [inline app/models/attorney.rb]: I'm not sure we need to stipulate "excluding judges" in these methods. If I'm looking to find all attorneys, I would already expect it not to return judges. I think we can leave these as `Attorney.list_all` and `AttorneyRepository.find_all_attorneys`. The logic for this was originally incorrect (by also including judges) and probably should have been named "find_all_attorneys_and_judges"

File: app/repositories/judge_repository.rb [hunk 5]
```diff
@@ -20,8 +20,19 @@ def self.find_all_judges_with_name_and_id
     end
   end
 
+  # includes acting judges
```
Comment (hschallhorn) [inline app/repositories/judge_repository.rb]: A+ to the clarifying comments

File: app/repositories/attorney_repository.rb (hunk 2 above)
Comment (yoomlam) [reply app/repositories/attorney_repository.rb]: No but I found it useful in understanding how data is modeled in VACOLS regarding the roles of attorneys, acting judges, and judges.

File: app/repositories/judge_repository.rb (hunk 3 above)
Comment (yoomlam) [reply app/repositories/judge_repository.rb]: They're not used. They are there for completeness and comprehension. Without them, one would have to go look up what 'J' and 'A' meant on line 25.

Comment (yoomlam) [reply app/repositories/judge_repository.rb]: I can convert them to comments if you prefer.

Comment (hschallhorn) [reply app/repositories/judge_repository.rb]: Yeah I think a comment above `judge_records` about what `svlj: %w[J A]` would be perfecto

**Comment by yoomlam** [comment]
@jimruggiero @araposo-tistatech This PR should be merged today 🤞 and deployed tomorrow morning (11am ET), unless I hear otherwise.

**Pull request merged**